Yda 5077 eus to python

stsnel opened the pull request:
Part of a pair of PRs, together with https://github.com/UtrechtUniversity/yoda/pull/234

stsnel committed:
YDA-5077: Set up core of Flask application

Set up core of new EUS Flask application, along with test
framework and Github Actions workflow.

stsnel committed:
Add parameter to disabled CSRF tokens

It is only meant to be disabled during integration tests, not during
normal use.

stsnel committed:
Extract function for generating new hash

stsnel committed:
add user: ensure changes are committed to db

stsnel committed:
Forgot password endpoint: initial implementation

stsnel committed:
Account activation: initial implementation

stsnel committed:
check auth: add integration test

The activation integration tests do not check authentication for
nonexistent accounts, so add a separate integration test for that.

stsnel committed:
Migrate reset password to Python

stsnel committed:
Remove old PHP code

stsnel committed:
whitespace

stsnel committed:
load test data: default to false

stsnel committed:
Fix encoding issues

Need to convert to/from UTF8 when performing bcrypt operations.

stsnel committed:
base template: fix missing block

stsnel committed:
Handle disabled CSRF in web templates

stsnel committed:
Various email fixes for Python EUS

stsnel committed:
Fix hash length to fit DB schema

stsnel committed:
Force API endpoint input to JSON

Ruleset doesn't set proper content type, so we need to force.

stsnel committed:
Fix up templates and form input params

stsnel committed:
Fix authentication header name

stsnel committed:
Fix encoding input values

stsnel committed:
Fix email recipient add confirmation

stsnel committed:
Typo fixes in mail template names

stsnel committed:
Fix layout activation web template

stsnel committed:
fix template params reset form

stsnel committed:
Show error message on wrong/missing API secret

stsnel committed:
Add option to disable API on web interface

This is used to disable the API on the main VHost, so that it is not
accessible from Internet.

stsnel committed:
Handle password whitespace more robustly

Password authentication failed in some circumstances due to trailing 0 bytes and
line breaks. Remove those before checking the password.

stsnel committed:
Pass error messages incorrect hash to templates

stsnel committed:
Fix creator variable name

stsnel committed:
Prefill username on password reset

This makes it work in the same way as activation.

lwesterhof commented on requirements.txt:
```diff
@@ -0,0 +1,9 @@
+Flask-SQLAlchemy==3.0.3
+Flask-wtf==1.1.1
+psycopg2-binary==2.9.5
+pytest-flask==1.2.0
```
We could move the test requirements to setup.py (https://setuptools.pypa.io/en/latest/userguide/dependency_management.html#optional-dependencies ):

```python
extras_require={
        "test": ["pytest-flask==1.2.0", "pytest==7.2.1"],
},
```

Installing the requirements:
`python -m pip install .[test]`

Then we do not have to manage the requirements in setup.py and requirements.txt.

lwesterhof commented on yoda_eus/app.py:
```diff
@@ -0,0 +1,534 @@
+db = SQLAlchemy()
+
+
+class User(db.Model):
```
Can we add some documentation to these classes?

lwesterhof commented on yoda_eus/mail.py:
```diff
@@ -0,0 +1,110 @@
+import email
```
Missing copyright and license.

lwesterhof commented on yoda_eus/password_complexity.py:
```diff
@@ -0,0 +1,34 @@
+import string
```
Missing copyright and license.

lwesterhof commented on yoda_eus/tests/test_integration.py:
```diff
@@ -0,0 +1,254 @@
+import base64
```
Missing copyright and license.

lwesterhof commented on yoda_eus/tests/test_unit.py:
```diff
@@ -0,0 +1,45 @@
+import string
```
Missing copyright and license.

lwesterhof commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,56 @@
+
+      - name: Check static typing
+        run: |
+          mypy . --explicit-package-bases || true
```
Maybe nice to enable this check? We only need to add type hints to a couple of lines to make it pass.

lwesterhof commented on yoda_eus/password_complexity.py:
```diff
@@ -0,0 +1,34 @@
+import string
+
+
+"""Checks whether a password meets EUS password complexity requirements.
```
Documentation should be moved below function definition.

lwesterhof commented on yoda_eus/app.py:
```diff
@@ -0,0 +1,534 @@
+
+    @app.route('/api/user/auth-check', methods=['POST'])
+    @csrf_exempt
+    def auth_check() -> Response:
```
Can you add docstrings, at least to all the app routes?

lwesterhof reviewed (changes requested):
Looks very good!

stsnel committed:
Add more copyright headers

stsnel committed:
fix docstring format / location

stsnel committed:
Enable MyPy

Need to add ignores for the Flask-SQLAlchemy model declarations
as a workaround for limitations regarding inheritance from objects, see
https://github.com/python/mypy/issues/8603

stsnel committed:
Integrate test requirements in setup.py

So requirements.txt is no longer needed

stsnel committed:
Add more function / class documentation

lwesterhof reviewed (approved):
LGTM

stsnel committed:
Add front-end password validation

Add basic front-end validation and feedback on password complexity and
length requirements, so that user gets feedback immediately when choosing
a password at the activation and password reset pages.

stsnel committed:
Activation form layout tweaks

Move terms of use down, and separate from password fields.

stsnel committed:
whitespace
